### PR TITLE
Copter: Add SmartRTL failsafe action

### PR DESCRIPTION
**Pull request by oneTimePad**
Adds SmartRTL or RTL and SmartRTL or Land failsafe options for batt, throttle, and GCS failsafes
#7552

### CODE REVIEW

File: ArduCopter/events.cpp [hunk 1]
```diff
@@ -218,6 +230,31 @@ void Copter::set_mode_RTL_or_land_with_pause(mode_reason_t reason)
     }
 }
 
+// set_mode_SmartRTL_or_land_with_pause - sets mode to SMART_RTL if possible or LAND with 4 second delay before descent starts
+// this is always called from a failsafe so we trigger notification to pilot
+void Copter::set_mode_SmartRTL_or_land_with_pause(mode_reason_t reason)
+{
+    // attempt to switch to SMART_RTL, if this failed then switch to Land
+    if (!set_mode(SMART_RTL, reason)) {
+        set_mode_land_with_pause(reason);
```
Comment (Pedals2Paddles) [inline ArduCopter/events.cpp]: Add `gcs().send_text(MAV_SEVERITY_WARNING, "SmartRTL Unavailable, Using Land Mode")` before set_mode_land_with_pause

File: ArduCopter/events.cpp [hunk 2]
```diff
@@ -218,6 +230,31 @@ void Copter::set_mode_RTL_or_land_with_pause(mode_reason_t reason)
...
+}
+
+// set_mode_SmartRTL_or_RTL - sets mode to SMART_RTL if possible or RTL if possible or LAND with 4 second delay before descent starts
+// this is always called from a failsafe so we trigger notification to pilot
+void Copter::set_mode_SmartRTL_or_RTL(mode_reason_t reason)
+{
+    // attempt to switch to SmartRTL, if this failed then attempt to RTL
+    // if that fails, then land
+    if (!set_mode(SMART_RTL, reason)) {
+        set_mode_RTL_or_land_with_pause(reason);
```
Comment (Pedals2Paddles) [inline ArduCopter/events.cpp]: Add `gcs().send_text(MAV_SEVERITY_WARNING, "SmartRTL Unavailable, Trying RTL Mode")` before  set_mode_RTL_or_land_with_pause.

**Comment by Pedals2Paddles** [comment]
I added some suggestions for GCS notification. So if SmartRTL fails, it will send a notification that's trying something else.  May want to add a similar GCS notification to the existing RTL_OR_LAND function too.

Other than that suggestion, LGTM! Will be great to have.

**Comment by oneTimePad** [comment]
Thank you both! I'll make the GCS notification changes and then see if I can figure out the LED/Buzzer notification additions.

**Comment by Pedals2Paddles** [comment]
I think a smartRTL specific LED pattern would end up being overridden by the existing low battery or radio failsafe LED patterns.  And even then, it would only be useful for people with external LEDs.  Same general issue with the buzzer. It would be overridden by other things, and is often not even audible over the prop noise anyway.

I think there is value in LED and audio notification updates.  But IMO it's not something useful enough to bog down getting this PR in.  That seems like something that should be it's own wider ranging effort on it's own unrelated PR.  IMO.

**Comment by rmackay9** [comment]
Merged, thanks!

I made some mostly non-functional changes as part of merging it:

- changed parameter descriptions to use lower case for flight mode names
- when checking the failsafe response parameter, we should leave the final option as a simple "else".  This is important in the unlikely case that the user sets the parameter to an unsupported value like, "10".

thanks again for this!

**Pull request closed without merging**